napi-inl.h

```diff
@@ -4256,6 +4256,131 @@ inline void AsyncWorker::OnWorkComplete(Napi::Env /*env*/, napi_status status) {
 }
 
 #if (NAPI_VERSION > 3)
+////////////////////////////////////////////////////////////////////////////////
+// ThreadSafeFunctionEx<Context> class
+////////////////////////////////////////////////////////////////////////////////
+
+// static
+template <typename ContextType>
+template <typename ResourceString, typename Finalizer, typename FinalizerDataType>
+inline ThreadSafeFunctionEx<ContextType> ThreadSafeFunctionEx<ContextType>::New(napi_env env,
+                                                  const Function& callback,
+                                                  const Object& resource,
+                                                  ResourceString resourceName,
+                                                  size_t maxQueueSize,
+                                                  size_t initialThreadCount,
+                                                  ContextType* context,
+                                                  Finalizer finalizeCallback,
+                                                  FinalizerDataType* data,
+                                                  napi_threadsafe_function_call_js call_js_cb) {
```

Comment:
I believe it is worth to type the `napi_threadsafe_function_call_js` with `ContextType` and `DataType`.

Reply:
@legendecas I was thinking that too, however looking at https://github.com/nodejs/node-addon-api/issues/594#issuecomment-581502550 , there were complaints that the "wrapping" for CallJS was a performance hit... and I can't think of a way to type `napi_threadsafe_function_call_js` without making a wrapper.

Furthermore, I don't think `DataType` can be typed without a wrapper, because it's "declared" in `::New` but "used" in `::[Non]BlockingCall`, and not sure how to link the two together. The way it works in the existing TSFN is that the `napi_threadsafe_function_call_js` takes this [`CallbackWrapper`](https://github.com/nodejs/node-addon-api/blob/89e62a9154b6057ed086a227db25caf26ac8c92a/napi-inl.h#L4585) data which is appropriately typed. Without a wrapper, not sure how to accomplish...